src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/LdapAccessPolicyDAO.scala

```diff
@@ -64,20 +64,25 @@ class LdapAccessPolicyDAO(
         IO(getAttributes(r, Attr.authDomain).map(WorkbenchGroupName))
     }
 
-  private def unmarshalResource(results: Entry): Either[String, Resource] =
-    for {
+  private def unmarshalResource(results: Entry): IO[Either[String, Resource]] = {
+    val result = for {
       resourceTypeName <- getAttribute(results, Attr.resourceType).toRight(s"${Attr.resourceType} attribute missing")
       resourceId <- getAttribute(results, Attr.resourceId).toRight(s"${Attr.resourceId} attribute missing")
     } yield {
       val authDomain = getAttributes(results, Attr.authDomain).map(g => WorkbenchGroupName(g)).toSet
       Resource(ResourceTypeName(resourceTypeName), ResourceId(resourceId), authDomain)
     }
+    IO.pure(result)
```

Comment:
since you just want to fail on the first error, I think you can just `IO.fromEither(result)`. This way, you won't need line 81 - 84